File: nahomy-c2-backend/src/infrastructure/messaging/RabbitMQAdapter.ts
import amqp, { type Connection, type Channel, type ConsumeMessage } from 'amqplib'
import type { IMessageBus } from '../../application/ports/IMessageBus'
import type { IDomainEvent } from '../../domain/events/IDomainEvent'
import { RabbitMQConfig } from '../config/RabbitMQConfig'

interface HandlerEntry {
  eventName: string
  handler: (event: IDomainEvent) => Promise<void>
}

export class RabbitMQAdapter implements IMessageBus {
  private connection: Connection | null = null
  private channel: Channel | null = null
  private readonly handlers: HandlerEntry[] = []
  private readonly config: RabbitMQConfig

  constructor(config: RabbitMQConfig) {
    this.config = config
  }

  async connect(): Promise<void> {
    this.connection = await amqp.connect(this.config.url)
    this.channel = await this.connection.createChannel()

    await this.channel.assertExchange(this.config.exchangeName, 'topic', {
      durable: true,
    })

    await this.channel.assertQueue(this.config.intentionQueue, { durable: true })
    await this.channel.assertQueue(this.config.auditQueue, { durable: true })
    await this.channel.assertQueue(this.config.metricsQueue, { durable: true })

    await this.channel.bindQueue(
      this.config.intentionQueue,
      this.config.exchangeName,
      'c2.intention.*',
    )
    await this.channel.bindQueue(
      this.config.auditQueue,
      this.config.exchangeName,
      'c2.audit.*',
    )
    await this.channel.bindQueue(
      this.config.metricsQueue,
      this.config.exchangeName,
      'c2.metrics.*',
    )

    await this.channel.bindQueue(
      this.config.intentionQueue,
      this.config.exchangeName,
      'c2.command.*',
    )

    this.startConsuming(this.config.intentionQueue)
    this.startConsuming(this.config.auditQueue)
    this.startConsuming(this.config.metricsQueue)

    console.log('[RabbitMQ] Conectado y bindeado a exchanges/queues')
  }

  async disconnect(): Promise<void> {
    await this.channel?.close()
    await this.connection?.close()
    console.log('[RabbitMQ] Desconectado')
  }

  async publish(event: IDomainEvent): Promise<void> {
    if (!this.channel) throw new Error('RabbitMQ no conectado')
    const buffer = Buffer.from(JSON.stringify(event))
    this.channel.publish(this.config.exchangeName, event.eventName, buffer, {
      persistent: true,
      contentType: 'application/json',
    })
    console.log(`[RabbitMQ] Publicado: ${event.eventName} | ID: ${event.eventId}`)
  }

  async publishMany(events: IDomainEvent[]): Promise<void> {
    for (const event of events) {
      await this.publish(event)
    }
  }

  subscribe(eventName: string, handler: (event: IDomainEvent) => Promise<void>): void {
    this.handlers.push({ eventName, handler })
  }

  private async startConsuming(queue: string): Promise<void> {
    if (!this.channel) return
    await this.channel.consume(
      queue,
      async (msg: ConsumeMessage | null) => {
        if (!msg) return
        try {
          const event: IDomainEvent = JSON.parse(msg.content.toString())
          for (const entry of this.handlers) {
            if (event.eventName.startsWith(entry.eventName.replace('*', ''))) {
              await entry.handler(event)
            }
          }
          this.channel!.ack(msg)
        } catch (error) {
          console.error('[RabbitMQ] Error procesando mensaje:', error)
          this.channel!.nack(msg, false, true)
        }
      },
      { noAck: false },
    )
  }
}

File: packages/server/src/infrastructure/messaging/rabbitmq_adapter.py
import json
import asyncio
import aio_pika
from ...domain.events import DomainEvent
from ...application.ports import IMessageBus


class RabbitMQAdapter(IMessageBus):
    def __init__(self, url: str, exchange_name: str = "c2.events"):
        self._url = url
        self._exchange_name = exchange_name
        self._connection: aio_pika.RobustConnection | None = None
        self._channel: aio_pika.Channel | None = None
        self._exchange: aio_pika.Exchange | None = None
        self._handlers: dict[str, callable] = {}

    async def connect(self) -> None:
        self._connection = await aio_pika.connect_robust(self._url)
        self._channel = await self._connection.channel()
        self._exchange = await self._channel.declare_exchange(
            self._exchange_name, aio_pika.ExchangeType.TOPIC, durable=True
        )

        queues = {
            "c2.intention.queue": ["c2.intention.*", "c2.command.*"],
            "c2.audit.queue": ["c2.audit.*"],
            "c2.metrics.queue": ["c2.metrics.*"],
        }

        for queue_name, routing_keys in queues.items():
            queue = await self._channel.declare_queue(queue_name, durable=True)
            for key in routing_keys:
                await queue.bind(self._exchange, key)
            await queue.consume(self._on_message)

        print("[RabbitMQ] Conectado y bindeado")

    async def disconnect(self) -> None:
        if self._connection:
            await self._connection.close()
            print("[RabbitMQ] Desconectado")

    async def publish(self, event: DomainEvent) -> None:
        if not self._exchange:
            raise RuntimeError("RabbitMQ no conectado")
        payload = {
            "event_id": event.event_id,
            "event_name": event.event_name,
            "occurred_on": event.occurred_on.isoformat(),
            "payload": {k: v for k, v in event.__dict__.items() if k not in ("event_id", "occurred_on")},
        }
        message = aio_pika.Message(
            body=json.dumps(payload, default=str).encode(),
            delivery_mode=aio_pika.DeliveryMode.PERSISTENT,
            content_type="application/json",
        )
        await self._exchange.send(
            routing_key=event.event_name,
            message=message,
        )
        print(f"[RabbitMQ] Publicado: {event.event_name}")

    async def publish_many(self, events: list[DomainEvent]) -> None:
        for event in events:
            await self.publish(event)

    def subscribe(self, event_name: str, handler) -> None:
        self._handlers[event_name] = handler

    async def _on_message(self, message: aio_pika.IncomingMessage) -> None:
        async with message.process():
            try:
                data = json.loads(message.body)
                event_name = data.get("event_name", "")
                for pattern, handler in self._handlers.items():
                    if event_name.startswith(pattern.replace("*", "")):
                        await handler(data)
            except Exception as e:
                print(f"[RabbitMQ] Error procesando: {e}")
